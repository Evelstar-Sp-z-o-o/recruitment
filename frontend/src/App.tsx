import { FC } from 'react';

import Root from '@/src/pages/Root';
import AppProviders from '@/src/utils/providers/AppProviders';

import './styles/main.scss';

const App: FC = () => {
  return (
    <AppProviders>
      <Root />
    </AppProviders>
  );
};

export default App;
