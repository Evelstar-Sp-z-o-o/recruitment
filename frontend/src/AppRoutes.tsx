import { Route, Routes, useLocation } from 'react-router-dom';

import Sidebar from './components/Sidebar';
import MainLayout from './components/layout/MainLayout';
import PostsLayout from './components/layout/PostsLayout';
import CreatePost from './pages/CreatePost';
import Home from './pages/Home';
import UpdatePost from './pages/UpdatePost';

const AppRoutes = () => {
  const location = useLocation();
  const background = location.state && location.state.background;

  return (
    <div className="main">
      <Sidebar />
      <Routes location={background || location}>
        <Route path="/" element={<MainLayout />}>
          <Route index element={<Home />} />
          <Route path="/posts" element={<PostsLayout />}>
            <Route path="update/:postId" element={<UpdatePost />} />
          </Route>
        </Route>
      </Routes>
      {background && (
        <Routes>
          <Route path="create" element={<CreatePost />} />
        </Routes>
      )}
    </div>
  );
};

export default AppRoutes;
