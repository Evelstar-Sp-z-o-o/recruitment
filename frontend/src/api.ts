import type { Post } from './types';

export const BASE_API_URL = 'http://localhost:3000/api';

const getPosts = async () => {
  const ENDPOINT = '/posts';

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Posts list.');
  }

  const data = await res.json();

  return data as unknown as Post[];
};

const getSinglePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Post details.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const deletePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'DELETE',
  });

  if (!res.ok) {
    throw new Error('Something went wrong when deleting post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const updatePost = async (postId: string, postData: Partial<Post>) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when updating post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const createPost = async (postData: Partial<Post>) => {
  const ENDPOINT = `/posts`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when creating post.');
  }

  const data = await res.json();

  return data as unknown as Post;
};

export { getPosts, getSinglePost, updatePost, createPost, deletePost };
