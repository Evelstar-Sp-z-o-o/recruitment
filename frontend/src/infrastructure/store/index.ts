import { useDispatch, useSelector } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';

import apiService from './services/apiService';
import userSlice from './slices/user';

const store = configureStore({
  reducer: {
    [apiService.reducerPath]: apiService.reducer,
    user: userSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiService.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector = useSelector.withTypes<RootState>();

export default store;
