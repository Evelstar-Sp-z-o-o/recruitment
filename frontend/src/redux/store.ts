import { configureStore } from '@reduxjs/toolkit';

import responseModalReducer from './responseModalSlice';

const store = configureStore({
  reducer: {
    responseModal: responseModalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
