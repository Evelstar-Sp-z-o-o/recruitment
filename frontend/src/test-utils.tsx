import { ReactNode } from 'react';

import AppProviders from '@/src/utils/providers/AppProviders';
import { render } from '@testing-library/react';

const AllTheProviders = ({ children }: { children: ReactNode }) => {
  return <AppProviders>{children}</AppProviders>;
};

const customRender = (ui, options?) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
