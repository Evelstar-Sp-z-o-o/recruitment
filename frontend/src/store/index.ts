import { postsApi } from '@/src/store/api/posts';
import { configureStore, createSlice } from '@reduxjs/toolkit';

const initialUser: null | string = null;
const initialLogin = false;

const userSlice = createSlice({
  name: 'user',
  initialState: initialUser,
  reducers: {
    setUser(_state, action) {
      return action.payload;
    },
  },
});

const loginSlice = createSlice({
  name: 'login',
  initialState: initialLogin,
  reducers: {
    setLogin(_state, action) {
      return action.payload;
    },
  },
});

export const { setUser } = userSlice.actions;
export const { setLogin } = loginSlice.actions;

export * from './api/posts.ts';

export const store = configureStore({
  reducer: {
    [postsApi.reducerPath]: postsApi.reducer,
    user: userSlice.reducer,
    login: loginSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(postsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
