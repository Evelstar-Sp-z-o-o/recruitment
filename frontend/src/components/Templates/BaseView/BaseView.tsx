import { FC, useState } from 'react';
import { useTranslation } from 'react-i18next';

import AddNewPost from '@/src/components/Atoms/AddNewPost/AddNewPost';
import Menu from '@/src/components/Molecules/Menu/Menu';
import PostModal from '@/src/components/Molecules/PostModal/PostModal';
import Footer from '@/src/components/Organisms/Footer/Footer';
import Header from '@/src/components/Organisms/Header/Header';
import PostsList from '@/src/components/Organisms/PostsList/PostsList';
import { IPost } from '@/src/store';
import { Alert, Snackbar } from '@mui/material';
import Typography from '@mui/material/Typography';
import { SerializedError } from '@reduxjs/toolkit';
import { FetchBaseQueryError } from '@reduxjs/toolkit/query';

const BaseView: FC = ({ posts }: { posts: IPost[] }) => {
  const [open, setOpen] = useState<boolean>(false);
  const [openPostModal, setOpenPostModal] = useState<boolean>(false);
  const [isPostSent, setIsPostSent] = useState<null | boolean>(null);
  const [openSnackbar, setOpenSnackbar] = useState<boolean>(false);
  const [singlePost, setSinglePost] = useState<null | IPost>(null);
  const { t } = useTranslation();

  const toggleMenu = (newOpen: boolean) => () => {
    setOpen(newOpen);
  };

  const handleTogglePostModal = () => {
    setSinglePost(null);
    setOpenPostModal((prevState) => !prevState);
  };

  const handleResponse = (res: FetchBaseQueryError | SerializedError | boolean) => {
    if (res) {
      setIsPostSent(typeof res === 'boolean');
      setOpenSnackbar(true);
    }
  };

  const handleSnackbar = () => {
    setOpenSnackbar(false);
  };

  const handleEdit = (post: IPost) => {
    setSinglePost(post);
    setOpenPostModal((prevState) => !prevState);
  };

  return (
    <>
      <Header toggleMenu={toggleMenu} />
      {posts && posts.length > 0 ? (
        <PostsList posts={posts} handleEdit={handleEdit} />
      ) : (
        <Typography color="text.secondary" align="center" sx={{ mt: '10rem', fontSize: '2rem' }}>
          {t('noPosts')}
        </Typography>
      )}
      <Menu
        open={open}
        toggleMenu={toggleMenu(false)}
        handleResponse={handleResponse}
        postsCount={posts?.length ?? 0}
      />
      <AddNewPost isFixed onClick={handleTogglePostModal} />
      <PostModal
        open={openPostModal}
        close={handleTogglePostModal}
        response={handleResponse}
        initialPost={singlePost}
        postsCount={posts?.length ?? 0}
      />
      <Snackbar open={openSnackbar} autoHideDuration={6000}>
        <Alert
          onClose={handleSnackbar}
          severity={isPostSent ? 'success' : 'error'}
          variant="filled"
          sx={{ width: '100%' }}
        >
          {isPostSent ? t('create.alert.success') : t('create.alert.error')}
        </Alert>
      </Snackbar>
      <Footer />
    </>
  );
};

export default BaseView;
