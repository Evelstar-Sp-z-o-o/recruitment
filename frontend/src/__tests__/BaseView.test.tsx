import React from 'react';

import BaseView from '@/src/components/Templates/BaseView/BaseView';
import { IPost, setUser, store } from '@/src/store';
import { render, screen, fireEvent } from '@/src/test-utils';
import '@testing-library/jest-dom';

const mockPosts: IPost[] = [
  {
    id: 1,
    data: {
      author: 'test@example.com',
      body: 'Test post content 1',
      created: 1722025907990,
      edited: 1722025907990,
    },
  },
  {
    id: 2,
    data: {
      author: 'test2@example.com',
      body: 'Test post content 2',
      created: 1722025907990,
      edited: 1722025907990,
    },
  },
];

beforeEach(() => {
  store.dispatch(setUser('test@example.com'));
});

describe('BaseView', () => {
  vi.mock('react-i18next', async (importOriginal) => {
    const actual = await importOriginal();
    return {
      ...actual,
      useTranslation: () => ({
        t: (key: string) => key,
      }),
      I18nextProvider: ({ children }) => <div>{children}</div>,
    };
  });

  it('renders posts when there are posts', () => {
    render(<BaseView posts={mockPosts} />);

    const postItems = screen.getAllByTestId(/EditIcon/);
    expect(postItems).toHaveLength(mockPosts.length);
  });

  it('renders no posts message when there are no posts', () => {
    render(<BaseView posts={[]} />);

    expect(screen.getByText('noPosts')).toBeInTheDocument();
  });

  it('calls handleEdit when the edit button is clicked', () => {
    render(<BaseView posts={mockPosts} />);

    const editButtons = screen.getAllByTestId(/EditIcon/);
    fireEvent.click(editButtons[0]);

    expect(screen.getByText('create.header')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<BaseView posts={mockPosts} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
