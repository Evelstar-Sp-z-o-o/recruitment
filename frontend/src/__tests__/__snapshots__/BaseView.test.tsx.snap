// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseView > should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionStatic css-owe5nn-MuiPaper-root-MuiAppBar-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-x6g243-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1qjb60v-MuiAvatar-root"
        >
          <img
            alt="User"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/src/assets/userIcon.svg"
          />
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <svg
        aria-hidden="true"
        aria-label="logo"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1o780gf-MuiSvgIcon-root"
        focusable="false"
      >
        <svg
          class="logo"
          fill="none"
          viewBox="0 0 325 325"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M107.5 92.5H53L107.5 38M107.5 92.5H150.5H162M107.5 92.5L53 38H107.5M107.5 92.5L162 38M107.5 38L162 92.5M107.5 38H162M162 92.5L216.5 38M162 92.5H216.5M216.5 38L271 92.5H216.5M216.5 38H162M216.5 38H271L216.5 92.5M162 38L216.5 92.5"
            stroke="black"
            stroke-width="16"
          />
          <path
            d="M162 100.5C162 100.5 162 195.5 162 251C162 287.031 117 285.5 117 285.5"
            stroke="black"
            stroke-width="45"
          />
        </svg>
      </svg>
    </header>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1x3n8rb-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-4vb5ob-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="TextsmsOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12zM7 9h2v2H7zm4 0h2v2h-2zm4 0h2v2h-2z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
          >
            test@example.com
          </span>
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
          >
            26.07.2024, 22:31
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
        >
          Test post content 1
        </p>
      </div>
      <div
        class="MuiCardActions-root css-5rkuwp-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteForeverIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1x3n8rb-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-4vb5ob-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="TextsmsOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12zM7 9h2v2H7zm4 0h2v2h-2zm4 0h2v2h-2z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
          >
            test2@example.com
          </span>
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
          >
            26.07.2024, 22:31
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
        >
          Test post content 2
        </p>
      </div>
      <div
        class="MuiCardActions-root css-5rkuwp-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </button>
        <button
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteForeverIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      aria-label="addPost"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1t609me-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element="true"
      role="button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-16960id-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <footer
      class="MuiContainer-root MuiContainer-maxWidthLg footer css-7f8y8b-MuiContainer-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 appName css-ahj2mt-MuiTypography-root"
      >
        TWIXER
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 copyrights css-ahj2mt-MuiTypography-root"
      >
        Â© 2024
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
